pkg/streamingpromql/operators/binary_operation_test.go

```diff
@@ -298,6 +299,231 @@ func TestBinaryOperation_SeriesMerging(t *testing.T) {
+					},
+					Histograms: []promql.HPoint{
+						{T: 3, H: &histogram.FloatHistogram{}},
+					},
+				},
+			},
+			sourceSeriesIndices: []int{0, 1},
+			sourceSeriesMetadata: []types.SeriesMetadata{
+				{Labels: labels.FromStrings("__name__", "right_side", "env", "test", "pod", "a")},
+				{Labels: labels.FromStrings("__name__", "right_side", "env", "test", "pod", "b")},
+			},
+			expectedOutput: types.InstantVectorSeriesData{
+				Floats: []promql.FPoint{
+					{T: 1, F: 10},
+					{T: 2, F: 20},
+					{T: 4, F: 40},
+					{T: 5, F: 50},
+				},
+				Histograms: []promql.HPoint{
+					{T: 3, H: &histogram.FloatHistogram{}},
+					{T: 6, H: &histogram.FloatHistogram{}},
+				},
+			},
+		},
+		"input series with both floats and histograms, series in conflict on different types": {
+			input: []types.InstantVectorSeriesData{
+				{
+					Floats: []promql.FPoint{
+						{T: 2, F: 20},
+					},
+				},
+				{
+					Histograms: []promql.HPoint{
+						{T: 2, H: &histogram.FloatHistogram{}},
+					},
+				},
+			},
+			sourceSeriesIndices: []int{0, 1},
+			sourceSeriesMetadata: []types.SeriesMetadata{
+				{Labels: labels.FromStrings("__name__", "right_side", "env", "test", "pod", "a")},
+				{Labels: labels.FromStrings("__name__", "right_side", "env", "test", "pod", "b")},
+			},
+			expectedError: `found both float and histogram samples for the match group FIXME on the right side of the operation at timestamp 1970-01-01T00:00:00.002Z`,
+		},
+		"input series with both floats and histograms, series in conflict on different type and not in time order": {
+			input: []types.InstantVectorSeriesData{
+				{
+					Histograms: []promql.HPoint{
+						{T: 3, H: &histogram.FloatHistogram{}},
+					},
+				},
+				{
+					Floats: []promql.FPoint{
+						{T: 1, F: 10},
+						{T: 3, F: 30},
+					},
+				},
+			},
+			sourceSeriesIndices: []int{0, 1},
+			sourceSeriesMetadata: []types.SeriesMetadata{
+				{Labels: labels.FromStrings("__name__", "right_side", "env", "test", "pod", "a")},
+				{Labels: labels.FromStrings("__name__", "right_side", "env", "test", "pod", "b")},
+			},
+			expectedError: `found both float and histogram samples for the match group FIXME on the right side of the operation at timestamp 1970-01-01T00:00:00.003Z`,
+		},
+		"input series have no points": {
+			input: []types.InstantVectorSeriesData{
+				{
+					Floats:     []promql.FPoint{},
+					Histograms: []promql.HPoint{},
+				},
+				{
+					Floats:     []promql.FPoint{},
+					Histograms: []promql.HPoint{},
+				},
+			},
+			sourceSeriesIndices: []int{0, 1},
+			sourceSeriesMetadata: []types.SeriesMetadata{
+				{Labels: labels.FromStrings("__name__", "right_side", "env", "test", "pod", "a")},
+				{Labels: labels.FromStrings("__name__", "right_side", "env", "test", "pod", "b")},
+			},
+			expectedOutput: types.InstantVectorSeriesData{
+				Floats:     nil,
+				Histograms: nil,
+			},
+		},
+		"input series overlap in different ways for floats and histograms": {
```

Comment:
It'd be good to add another test case for the scenario where some input series have floats and no histograms, some have just histograms, and some have both.